File: AutoRest/Generators/Java/Azure.Java.Tests/src/test/java/fixtures/resourceflattening/ResourceFlatteningTests.java
package fixtures.resourceflattening;

import com.microsoft.rest.ServiceResponse;
import com.microsoft.azure.serializer.AzureJacksonMapperAdapter;
import com.squareup.okhttp.OkHttpClient;
import fixtures.resourceflattening.models.FlattenedProduct;
import fixtures.resourceflattening.models.Resource;
import fixtures.resourceflattening.models.ResourceCollection;
import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;
import retrofit.JacksonConverterFactory;
import retrofit.Retrofit;

import java.net.CookieManager;
import java.net.CookiePolicy;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class ResourceFlatteningTests {
    private static AutoRestResourceFlatteningTestService client;

    @BeforeClass
    public static void setup() {
        OkHttpClient httpClient = new OkHttpClient();
        CookieManager cookieManager = new CookieManager();
        cookieManager.setCookiePolicy(CookiePolicy.ACCEPT_ALL);
        httpClient.setCookieHandler(cookieManager);
        Executor executor = Executors.newCachedThreadPool();
        Retrofit.Builder builder = new Retrofit.Builder()
                .addConverterFactory(JacksonConverterFactory.create(new AzureJacksonMapperAdapter().getObjectMapper()))
                .callbackExecutor(executor);

        client = new AutoRestResourceFlatteningTestServiceImpl("http://localhost.:3000", null, httpClient, builder);
        client.setLongRunningOperationRetryTimeout(0);
    }

    @Test
    public void getArray() throws Exception {
        List<FlattenedProduct> result = client.getArray().getBody();
        Assert.assertEquals(3, result.size());
        // Resource 1
        Assert.assertEquals("1", result.get(0).getId());
        Assert.assertEquals("OK", result.get(0).getProvisioningStateValues());
        Assert.assertEquals("Product1", result.get(0).getPname());
        Assert.assertEquals("Flat", result.get(0).getFlattenedProductType());
        Assert.assertEquals("Building 44", result.get(0).getLocation());
        Assert.assertEquals("Resource1", result.get(0).getName());
        Assert.assertEquals("Succeeded", result.get(0).getProvisioningState());
        Assert.assertEquals("Microsoft.Web/sites", result.get(0).getType());
        Assert.assertEquals("value1", result.get(0).getTags().get("tag1"));
        Assert.assertEquals("value3", result.get(0).getTags().get("tag2"));
        // Resource 2
        Assert.assertEquals("2", result.get(1).getId());
        Assert.assertEquals("Resource2", result.get(1).getName());
        Assert.assertEquals("Building 44", result.get(1).getLocation());
        // Resource 3
        Assert.assertEquals("3", result.get(2).getId());
        Assert.assertEquals("Resource3", result.get(2).getName());
    }

    @Test
    public void putArray() throws Exception {
        List<Resource> body = new ArrayList<>();
        FlattenedProduct product = new FlattenedProduct();
        product.setLocation("West US");
        product.setTags(new HashMap<String, String>());
        product.getTags().put("tag1", "value1");
        product.getTags().put("tag2", "value3");
        body.add(product);
        FlattenedProduct product1 = new FlattenedProduct();
        product1.setLocation("Building 44");
        body.add(product1);
        ServiceResponse<Void> response = client.putArray(body);
        Assert.assertEquals(200, response.getResponse().code());
    }

    @Test
    public void getDictionary() throws Exception {
        Map<String, FlattenedProduct> result = client.getDictionary().getBody();
        Assert.assertEquals(3, result.size());
        // Resource 1
        Assert.assertEquals("1", result.get("Product1").getId());
        Assert.assertEquals("OK", result.get("Product1").getProvisioningStateValues());
        Assert.assertEquals("Product1", result.get("Product1").getPname());
        Assert.assertEquals("Flat", result.get("Product1").getFlattenedProductType());
        Assert.assertEquals("Building 44", result.get("Product1").getLocation());
        Assert.assertEquals("Resource1", result.get("Product1").getName());
        Assert.assertEquals("Succeeded", result.get("Product1").getProvisioningState());
        Assert.assertEquals("Microsoft.Web/sites", result.get("Product1").getType());
        Assert.assertEquals("value1", result.get("Product1").getTags().get("tag1"));
        Assert.assertEquals("value3", result.get("Product1").getTags().get("tag2"));
        // Resource 2
        Assert.assertEquals("2", result.get("Product2").getId());
        Assert.assertEquals("Resource2", result.get("Product2").getName());
        Assert.assertEquals("Building 44", result.get("Product2").getLocation());
        // Resource 3
        Assert.assertEquals("3", result.get("Product3").getId());
        Assert.assertEquals("Resource3", result.get("Product3").getName());
    }

    @Test
    public void putDictionary() throws Exception {
        Map<String, FlattenedProduct> body = new HashMap<>();
        FlattenedProduct product = new FlattenedProduct();
        product.setLocation("West US");
        product.setTags(new HashMap<String, String>());
        product.getTags().put("tag1", "value1");
        product.getTags().put("tag2", "value3");
        product.setPname("Product1");
        product.setFlattenedProductType("Flat");
        body.put("Resource1", product);
        FlattenedProduct product1 = new FlattenedProduct();
        product1.setLocation("Building 44");
        product1.setPname("Product2");
        product1.setFlattenedProductType("Flat");
        body.put("Resource2", product1);
        ServiceResponse<Void> response = client.putDictionary(body);
        Assert.assertEquals(200, response.getResponse().code());
    }

    @Test
    public void getResourceCollection() throws Exception {
        ResourceCollection resultResource = client.getResourceCollection().getBody();
        //Dictionaryofresources
        Assert.assertEquals(3, resultResource.getDictionaryofresources().size());
        // Resource 1
        Assert.assertEquals("1", resultResource.getDictionaryofresources().get("Product1").getId());
        Assert.assertEquals("OK", resultResource.getDictionaryofresources().get("Product1").getProvisioningStateValues());
        Assert.assertEquals("Product1", resultResource.getDictionaryofresources().get("Product1").getPname());
        Assert.assertEquals("Flat", resultResource.getDictionaryofresources().get("Product1").getFlattenedProductType());
        Assert.assertEquals("Building 44", resultResource.getDictionaryofresources().get("Product1").getLocation());
        Assert.assertEquals("Resource1", resultResource.getDictionaryofresources().get("Product1").getName());
        Assert.assertEquals("Succeeded", resultResource.getDictionaryofresources().get("Product1").getProvisioningState());
        Assert.assertEquals("Microsoft.Web/sites", resultResource.getDictionaryofresources().get("Product1").getType());
        Assert.assertEquals("value1", resultResource.getDictionaryofresources().get("Product1").getTags().get("tag1"));
        Assert.assertEquals("value3", resultResource.getDictionaryofresources().get("Product1").getTags().get("tag2"));
        // Resource 2
        Assert.assertEquals("2", resultResource.getDictionaryofresources().get("Product2").getId());
        Assert.assertEquals("Resource2", resultResource.getDictionaryofresources().get("Product2").getName());
        Assert.assertEquals("Building 44", resultResource.getDictionaryofresources().get("Product2").getLocation());
        // Resource 3
        Assert.assertEquals("3", resultResource.getDictionaryofresources().get("Product3").getId());
        Assert.assertEquals("Resource3", resultResource.getDictionaryofresources().get("Product3").getName());

        //Arrayofresources
        Assert.assertEquals(3, resultResource.getArrayofresources().size());
        // Resource 1
        Assert.assertEquals("4", resultResource.getArrayofresources().get(0).getId());
        Assert.assertEquals("OK", resultResource.getArrayofresources().get(0).getProvisioningStateValues());
        Assert.assertEquals("Product4", resultResource.getArrayofresources().get(0).getPname());
        Assert.assertEquals("Flat", resultResource.getArrayofresources().get(0).getFlattenedProductType());
        Assert.assertEquals("Building 44", resultResource.getArrayofresources().get(0).getLocation());
        Assert.assertEquals("Resource4", resultResource.getArrayofresources().get(0).getName());
        Assert.assertEquals("Succeeded", resultResource.getArrayofresources().get(0).getProvisioningState());
        Assert.assertEquals("Microsoft.Web/sites", resultResource.getArrayofresources().get(0).getType());
        Assert.assertEquals("value1", resultResource.getArrayofresources().get(0).getTags().get("tag1"));
        Assert.assertEquals("value3", resultResource.getArrayofresources().get(0).getTags().get("tag2"));
        // Resource 2
        Assert.assertEquals("5", resultResource.getArrayofresources().get(1).getId());
        Assert.assertEquals("Resource5", resultResource.getArrayofresources().get(1).getName());
        Assert.assertEquals("Building 44", resultResource.getArrayofresources().get(1).getLocation());
        // Resource 3
        Assert.assertEquals("6", resultResource.getArrayofresources().get(2).getId());
        Assert.assertEquals("Resource6", resultResource.getArrayofresources().get(2).getName());

        //productresource
        Assert.assertEquals("7", resultResource.getProductresource().getId());
        Assert.assertEquals("Resource7", resultResource.getProductresource().getName());
    }

    @Test
    public void putResourceCollection() throws Exception {
        Map<String, FlattenedProduct> resources = new HashMap<>();
        resources.put("Resource1", new FlattenedProduct());
        resources.get("Resource1").setLocation("West US");
        resources.get("Resource1").setPname("Product1");
        resources.get("Resource1").setFlattenedProductType("Flat");
        resources.get("Resource1").setTags(new HashMap<String, String>());
        resources.get("Resource1").getTags().put("tag1", "value1");
        resources.get("Resource1").getTags().put("tag2", "value3");

        resources.put("Resource2", new FlattenedProduct());
        resources.get("Resource2").setLocation("Building 44");
        resources.get("Resource2").setPname("Product2");
        resources.get("Resource2").setFlattenedProductType("Flat");

        ResourceCollection complexObj = new ResourceCollection();
        complexObj.setDictionaryofresources(resources);
        complexObj.setArrayofresources(new ArrayList<FlattenedProduct>());
        complexObj.getArrayofresources().add(resources.get("Resource1"));
        FlattenedProduct p1 = new FlattenedProduct();
        p1.setLocation("East US");
        p1.setPname("Product2");
        p1.setFlattenedProductType("Flat");
        complexObj.getArrayofresources().add(p1);
        FlattenedProduct pr = new FlattenedProduct();
        pr.setLocation("India");
        pr.setPname("Azure");
        pr.setFlattenedProductType("Flat");
        complexObj.setProductresource(pr);

        ServiceResponse<Void> response = client.putResourceCollection(complexObj);
        Assert.assertEquals(200, response.getResponse().code());
    }
}

File: AutoRest/Generators/Python/Azure.Python.Tests/AcceptanceTests/resource_flattening_tests.py
import unittest
import subprocess
import sys
import isodate
import tempfile
import json
from uuid import uuid4
from datetime import date, datetime, timedelta
import os
from os.path import dirname, pardir, join, realpath, sep, pardir

cwd = dirname(realpath(__file__))
root = realpath(join(cwd , pardir, pardir, pardir, pardir, pardir))
sys.path.append(join(root, "ClientRuntimes" , "Python", "msrest"))
sys.path.append(join(root, "ClientRuntimes" , "Python", "msrestazure"))

tests = realpath(join(cwd, pardir, "Expected", "AcceptanceTests"))
sys.path.append(join(tests, "ResourceFlattening"))

from msrest.serialization import Deserializer
from msrest.exceptions import DeserializationError
from msrest.authentication import BasicTokenAuthentication

from auto_rest_resource_flattening_test_service import AutoRestResourceFlatteningTestService, AutoRestResourceFlatteningTestServiceConfiguration
from auto_rest_resource_flattening_test_service.models import FlattenedProduct, ErrorException, ResourceCollection


class ResourceFlatteningTests(unittest.TestCase):

    def setUp(self):
        cred = BasicTokenAuthentication({"access_token" :str(uuid4())})
        config = AutoRestResourceFlatteningTestServiceConfiguration(cred, base_url="http://localhost:3000")
        config.log_level = 10
        self.client = AutoRestResourceFlatteningTestService(config)

        return super(ResourceFlatteningTests, self).setUp()

    def test_flattening_array(self):

        #Array
        result = self.client.get_array()
        self.assertEqual(3, len(result))
        # Resource 1
        self.assertEqual("1", result[0].id)
        self.assertEqual("OK", result[0].provisioning_state_values)
        self.assertEqual("Product1", result[0].pname)
        self.assertEqual("Flat", result[0].flattened_product_type)
        self.assertEqual("Building 44", result[0].location)
        self.assertEqual("Resource1", result[0].name)
        self.assertEqual("Succeeded", result[0].provisioning_state)
        self.assertEqual("Microsoft.Web/sites", result[0].type)
        self.assertEqual("value1", result[0].tags["tag1"])
        self.assertEqual("value3", result[0].tags["tag2"])
        # Resource 2
        self.assertEqual("2", result[1].id)
        self.assertEqual("Resource2", result[1].name)
        self.assertEqual("Building 44", result[1].location)
        # Resource 3
        self.assertEqual("3", result[2].id)
        self.assertEqual("Resource3", result[2].name)

        resourceArray = [
                FlattenedProduct(
                    location = "West US",
                    tags = {"tag1":"value1", "tag2":"value3"}),
                FlattenedProduct(
                    location = "Building 44")]

        self.client.put_array(resourceArray)
        pass

    def test_flattening_dictionary(self):

        #Dictionary
        resultDictionary = self.client.get_dictionary()
        self.assertEqual(3, len(resultDictionary))
        # Resource 1
        self.assertEqual("1", resultDictionary["Product1"].id)
        self.assertEqual("OK", resultDictionary["Product1"].provisioning_state_values)
        self.assertEqual("Product1", resultDictionary["Product1"].pname)
        self.assertEqual("Flat", resultDictionary["Product1"].flattened_product_type)
        self.assertEqual("Building 44", resultDictionary["Product1"].location)
        self.assertEqual("Resource1", resultDictionary["Product1"].name)
        self.assertEqual("Succeeded", resultDictionary["Product1"].provisioning_state)
        self.assertEqual("Microsoft.Web/sites", resultDictionary["Product1"].type)
        self.assertEqual("value1", resultDictionary["Product1"].tags["tag1"])
        self.assertEqual("value3", resultDictionary["Product1"].tags["tag2"])
        # Resource 2
        self.assertEqual("2", resultDictionary["Product2"].id)
        self.assertEqual("Resource2", resultDictionary["Product2"].name)
        self.assertEqual("Building 44", resultDictionary["Product2"].location)
        # Resource 3
        self.assertEqual("3", resultDictionary["Product3"].id)
        self.assertEqual("Resource3", resultDictionary["Product3"].name)

        resourceDictionary = {
                "Resource1": FlattenedProduct(
                    location = "West US",
                    tags = {"tag1":"value1", "tag2":"value3"},
                    pname = "Product1",
                    flattened_product_type = "Flat"),
                "Resource2": FlattenedProduct(
                    location = "Building 44",
                    pname = "Product2",
                    flattened_product_type = "Flat")}
        #TODO: serializer need to handle flatten resources
        #self.client.put_dictionary(resourceDictionary)
        pass

    def test_flattening_complex_object(self):

        #ResourceCollection
        resultResource = self.client.get_resource_collection()

        #dictionaryofresources
        self.assertEqual(3, len(resultResource.dictionaryofresources))
        # Resource 1
        self.assertEqual("1", resultResource.dictionaryofresources["Product1"].id)
        self.assertEqual("OK", resultResource.dictionaryofresources["Product1"].provisioning_state_values)
        self.assertEqual("Product1", resultResource.dictionaryofresources["Product1"].pname)
        self.assertEqual("Flat", resultResource.dictionaryofresources["Product1"].flattened_product_type)
        self.assertEqual("Building 44", resultResource.dictionaryofresources["Product1"].location)
        self.assertEqual("Resource1", resultResource.dictionaryofresources["Product1"].name)
        self.assertEqual("Succeeded", resultResource.dictionaryofresources["Product1"].provisioning_state)
        self.assertEqual("Microsoft.Web/sites", resultResource.dictionaryofresources["Product1"].type)
        self.assertEqual("value1", resultResource.dictionaryofresources["Product1"].tags["tag1"])
        self.assertEqual("value3", resultResource.dictionaryofresources["Product1"].tags["tag2"])
        # Resource 2
        self.assertEqual("2", resultResource.dictionaryofresources["Product2"].id)
        self.assertEqual("Resource2", resultResource.dictionaryofresources["Product2"].name)
        self.assertEqual("Building 44", resultResource.dictionaryofresources["Product2"].location)
        # Resource 3
        self.assertEqual("3", resultResource.dictionaryofresources["Product3"].id)
        self.assertEqual("Resource3", resultResource.dictionaryofresources["Product3"].name)

        #arrayofresources
        self.assertEqual(3, len(resultResource.arrayofresources))
        # Resource 1
        self.assertEqual("4", resultResource.arrayofresources[0].id)
        self.assertEqual("OK", resultResource.arrayofresources[0].provisioning_state_values)
        self.assertEqual("Product4", resultResource.arrayofresources[0].pname)
        self.assertEqual("Flat", resultResource.arrayofresources[0].flattened_product_type)
        self.assertEqual("Building 44", resultResource.arrayofresources[0].location)
        self.assertEqual("Resource4", resultResource.arrayofresources[0].name)
        self.assertEqual("Succeeded", resultResource.arrayofresources[0].provisioning_state)
        self.assertEqual("Microsoft.Web/sites", resultResource.arrayofresources[0].type)
        self.assertEqual("value1", resultResource.arrayofresources[0].tags["tag1"])
        self.assertEqual("value3", resultResource.arrayofresources[0].tags["tag2"])
        # Resource 2
        self.assertEqual("5", resultResource.arrayofresources[1].id)
        self.assertEqual("Resource5", resultResource.arrayofresources[1].name)
        self.assertEqual("Building 44", resultResource.arrayofresources[1].location)
        # Resource 3
        self.assertEqual("6", resultResource.arrayofresources[2].id)
        self.assertEqual("Resource6", resultResource.arrayofresources[2].name)

        #productresource
        self.assertEqual("7", resultResource.productresource.id)
        self.assertEqual("Resource7", resultResource.productresource.name)

        resourceDictionary = {
                "Resource1": FlattenedProduct(
                    location = "West US",
                    tags = {"tag1":"value1", "tag2":"value3"},
                    pname = "Product1",
                    flattened_product_type = "Flat"),
                "Resource2": FlattenedProduct(
                    location = "Building 44",
                    pname = "Product2",
                    flattened_product_type = "Flat")}

        resourceComplexObject = ResourceCollection(
                dictionaryofresources = resourceDictionary,
                arrayofresources = [
                    FlattenedProduct(
                        location = "West US",
                        tags = {"tag1":"value1", "tag2":"value3"},
                        pname = "Product1",
                        flattened_product_type = "Flat"),
                    FlattenedProduct(
                        location = "East US",
                        pname = "Product2",
                        flattened_product_type = "Flat")],
                productresource = FlattenedProduct(
                    location = "India",
                    pname = "Azure",
                    flattened_product_type = "Flat"))
        #TODO: serializer need to handle flatten resources
        #self.client.put_resource_collection(resourceComplexObject)
        pass

if __name__ == '__main__':
    unittest.main()